Data/ByteString/Short/Internal.hs

```diff
@@ -446,6 +586,856 @@ concat sbss =
+--
+-- To skip to the first occurence of a string:
+--
+-- > snd (breakSubstring x y)
+--
+-- To take the parts of a string before a delimiter:
+--
+-- > fst (breakSubstring x y)
+--
+-- Note that calling `breakSubstring x` does some preprocessing work, so
+-- you should avoid unnecessarily duplicating breakSubstring calls with the same
+-- pattern.
+--
+-- @since 0.11.3.0
+breakSubstring :: ShortByteString -- ^ String to search for
+               -> ShortByteString -- ^ String to search in
+               -> (ShortByteString, ShortByteString) -- ^ Head and tail of string broken at substring
+breakSubstring pat =
+  case lp of
+    0 -> (mempty,)
+    1 -> breakByte (head pat)
+    _ -> if lp * 8 <= finiteBitSize (0 :: Word)
+             then shift
+             else karpRabin
+  where
+    lp = length pat
+    karpRabin :: ShortByteString -> (ShortByteString, ShortByteString)
+    karpRabin src
+        | length src < lp = (src,mempty)
+        | otherwise = search (rollingHash $ take lp src) lp
+      where
+        k           = 2891336453 :: Word32
+        rollingHash = foldl' (\h b -> h * k + fromIntegral b) 0
+        hp          = rollingHash pat
+        m           = k ^ lp
+        get = fromIntegral . unsafeIndex src
+        search !hs !i
+            | hp == hs && pat == take lp b = u
+            | length src <= i           = (src, mempty) -- not found
+            | otherwise                    = search hs' (i + 1)
+          where
+            u@(_, b) = splitAt (i - lp) src
+            hs' = hs * k +
+                  get i -
+                  m * get (i - lp)
+
+    shift :: ShortByteString -> (ShortByteString, ShortByteString)
+    shift !src
+        | length src < lp = (src, mempty)
+        | otherwise       = search (intoWord $ take lp src) lp
+      where
+        intoWord :: ShortByteString -> Word
+        intoWord = foldl' (\w b -> (w `shiftL` 8) .|. fromIntegral b) 0
+        wp   = intoWord pat
+        mask' = (1 `shiftL` (8 * lp)) - 1
+        search !w !i
+            | w == wp            = splitAt (i - lp) src
+            | length src <= i = (src, mempty)
+            | otherwise       = search w' (i + 1)
+          where
+            b  = fromIntegral (unsafeIndex src i)
+            w' = mask' .&. ((w `shiftL` 8) .|. b)
+
+
+-- --------------------------------------------------------------------
+-- Searching ShortByteString
+
+-- | /O(n)/ 'elem' is the 'ShortByteString' membership predicate.
+--
+-- @since 0.11.3.0
+elem :: Word8 -> ShortByteString -> Bool
+elem c = \sbs -> case elemIndex c sbs of Nothing -> False ; _ -> True
+
+-- | /O(n)/ 'filter', applied to a predicate and a ByteString,
+-- returns a ByteString containing those characters that satisfy the
+-- predicate.
+--
+-- @since 0.11.3.0
+filter :: (Word8 -> Bool) -> ShortByteString -> ShortByteString
+filter k = \sbs -> if
+    | null sbs  -> sbs
+    | otherwise -> pack . List.filter k . unpack $ sbs
```

Comment:
```
-- RHS size: {terms: 22, types: 9, coercions: 0, joins: 0/1}
Data.ByteString.Short.Internal.$wfilter [InlPrag=NOUSERINLINE[2]]
  :: (Word8 -> Bool) -> ByteArray# -> ShortByteString
[GblId,
 Arity=2,
 Caf=NoCafRefs,
 Str=<L,C(U)><L,U>,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True, Guidance=IF_ARGS [0 0] 161 20}]
Data.ByteString.Short.Internal.$wfilter
  = \ (w_sKaL :: Word8 -> Bool) (ww_sKaP :: ByteArray#) ->
      case sizeofByteArray# ww_sKaP of {
        __DEFAULT ->
          let {
            ws_sHgJ [Dmd=<S,U>] :: [Word8]
            [LclId]
            ws_sHgJ
              = List.filter
                  @ Word8
                  w_sKaL
                  (Data.ByteString.Short.Internal.$wunpack ww_sKaP) } in
          case GHC.List.$wlenAcc @ Word8 ws_sHgJ 0# of ww2_aiuF
          { __DEFAULT ->
          Data.ByteString.Short.Internal.$wpackLenBytes ww2_aiuF ws_sHgJ
          };
        0# -> Data.ByteString.Short.Internal.SBS ww_sKaP
      }

-- RHS size: {terms: 8, types: 5, coercions: 0, joins: 0/0}
filter [InlPrag=NOUSERINLINE[2]]
  :: (Word8 -> Bool) -> ShortByteString -> ShortByteString
[GblId,
 Arity=2,
 Caf=NoCafRefs,
 Str=<L,C(U)><S,1*U(U)>,
 Unf=Unf{Src=InlineStable, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=True,boring_ok=False)
         Tmpl= \ (w_sKaL [Occ=Once1] :: Word8 -> Bool)
                 (w1_sKaM [Occ=Once1!] :: ShortByteString) ->
                 case w1_sKaM of { SBS ww1_sKaP [Occ=Once1] ->
                 Data.ByteString.Short.Internal.$wfilter w_sKaL ww1_sKaP
                 }}]
filter
  = \ (w_sKaL :: Word8 -> Bool) (w1_sKaM :: ShortByteString) ->
      case w1_sKaM of { SBS ww1_sKaP ->
      Data.ByteString.Short.Internal.$wfilter w_sKaL ww1_sKaP
      }
```

Reply:
this is with `-ddump-simpl`

Reply:
Yeah, this looks like it would allocate quite a bit. Doing this with `createAndTrim` and without the intermediate list should be significantly more efficient.